src/tests/suite/daos_dist_tx.c

```diff
@@ -0,0 +1,986 @@
+	punch_akey(dkey, akey, th, &req);
+
+	req.arg->expect_result = 0;
+
+	lookup_single(dkey, akey, 0, fetch_buf, DTX_IO_SMALL, th, &req);
+	assert_int_equal(req.iod[0].iod_size, DTX_IO_SMALL);
+
+	lookup_single(dkey, akey2, 0, fetch_buf, DTX_IO_SMALL, th, &req);
+	assert_int_equal(req.iod[0].iod_size, 0);
+
+	MUST(daos_tx_commit(th, NULL));
+
+	ioreq_fini(&req);
+	MUST(daos_tx_close(th, NULL));
+}
+
+static void
+dtx_12(void **state)
+{
+	test_arg_t	*arg = *state;
+	const char	*dkey = dts_dtx_dkey;
+	const char	*akey = dts_dtx_akey;
+	const char	*akey2 = "tmp_akey";
+	char		 write_buf[DTX_IO_SMALL];
+	char		 fetch_buf[DTX_IO_SMALL];
+	daos_handle_t	 th = { 0 };
+	daos_obj_id_t	 oid;
+	struct ioreq	 req;
+	int		 i;
+
+	print_message("DTX with zero copy flag\n");
+	MUST(daos_tx_open(arg->coh, &th, DAOS_TF_ZERO_COPY, NULL));
+
+	arg->async = 0;
+	oid = dts_oid_gen(OC_RP_XSF, 0, arg->myrank);
+	ioreq_init(&req, arg->coh, oid, DAOS_IOD_ARRAY, arg);
+
+	dts_buf_render(write_buf, DTX_IO_SMALL);
+	insert_single(dkey, akey, 0, write_buf, DTX_IO_SMALL, th, &req);
+
+	/* Reuse 'write_buf' */
+	for (i = 0; i < DTX_IO_SMALL; i++)
+		write_buf[i] += 1;
+
+	insert_single(dkey, akey2, 0, write_buf, DTX_IO_SMALL, th, &req);
+
+	MUST(daos_tx_commit(th, NULL));
+
+	lookup_single(dkey, akey2, 0, fetch_buf, DTX_IO_SMALL,
+		      DAOS_TX_NONE, &req);
+	assert_memory_equal(write_buf, fetch_buf, DTX_IO_SMALL);
+
+	lookup_single(dkey, akey, 0, fetch_buf, DTX_IO_SMALL,
+		      DAOS_TX_NONE, &req);
+	/* 'write_buf' has been overwritten, so it is the same as 2nd update. */
+	assert_memory_equal(write_buf, fetch_buf, DTX_IO_SMALL);
+
+	ioreq_fini(&req);
+	MUST(daos_tx_close(th, NULL));
+}
+
+static void
+dtx_13(void **state)
+{
+	test_arg_t	*arg = *state;
+	const char	*dkey = dts_dtx_dkey;
+	const char	*akey = dts_dtx_akey;
+	char		 write_buf[DTX_IO_SMALL];
+	char		 fetch_buf[DTX_IO_SMALL];
+	daos_handle_t	 th = { 0 };
+	daos_obj_id_t	 oid;
+	struct ioreq	 req;
+	int		 rc;
+
+	print_message("DTX status machnie\n");
+
+	arg->async = 0;
+	oid = dts_oid_gen(OC_RP_XSF, 0, arg->myrank);
+	ioreq_init(&req, arg->coh, oid, DAOS_IOD_ARRAY, arg);
+	dts_buf_render(write_buf, DTX_IO_SMALL);
+
+	print_message("Open the TX1...\n");
+	MUST(daos_tx_open(arg->coh, &th, 0, NULL));
+
+	print_message("Commit the empty TX1...\n");
+	MUST(daos_tx_commit(th, NULL));
+
+	print_message("Commit the committed TX1...\n");
+	MUST(daos_tx_commit(th, NULL));
+
+	print_message("Abort the committed TX1...\n");
+	rc = daos_tx_abort(th, NULL);
+	assert_int_equal(rc, -DER_NO_PERM);
+
+	print_message("Restart the committed TX1...\n");
+	rc = daos_tx_restart(th, NULL);
+	assert_int_equal(rc, -DER_NO_PERM);
+
+	print_message("Update against the committed TX1...\n");
+	req.arg->expect_result = -DER_NO_PERM;
+	insert_single(dkey, akey, 0, write_buf, DTX_IO_SMALL, th, &req);
+
+	print_message("Fetch against the committed TX1...\n");
+	req.arg->expect_result = 0;
```

Comment:
Anytime, we should allow the user to fetch data against the TX handle, in spite of whether it is committed or not. The difference is that if the TX is not committed, the fetch operation cannot see TX itself modifications; otherwise, related committed modifications should be returned by the fetch. Then why need "NO_PERM" for such case?

Reply:
IMO a committed TX should not be used to read/write anymore on it.
from a user perspective this seems weird. why not use TX_NONE in that case?

Reply:
this is my opinion, but we can discuss this more on call

Reply:
> IMO a committed TX should not be used to read/write anymore on it.
> from a user perspective this seems weird. why not use TX_NONE in that case?

In the test, we can use 'tx_NONE'. But in the real world, there may be other TXs overwrite our commit value with newer epochs, so the user may has to use its own TX's epoch to fetch after the commit.

Reply:
but I don't see that really as a real use case. but more of a testing one.
If a user wants to read from a consistent TX, they should not be using a committed one but should create a snapshot and read from the snapshot instead.

Reply:
The user can verify its data against the snapshot, but the pre-condition is that it needs to create a snapshot just after its TX commit, that is inefficient than reading within the same TX. On the other hand, the user needs to know the TX's epoch when it try to open the snapshot. But we may cannot know the epoch until the TX is committed. So means that we still need to use the committed TX handle for such purpose.

Reply:
i just don't think this is a valid use case from an application perspective.

let's say someone else overwrite the data in another future tx, and commits.
then aggregation happens and data from original tx is aggregated.
what happens when the original tx with the older data is read with the committed tx when there is no snapshot?